dogechain-lab/dogechain
# [HardFork] portland hardfork on fixing bridge withdrawn issue

# Description

The `portland` hardfork should be merged to the `dev` branch now, since most of the community nodes are upgraded.

It was brought out by the `QT Wallet` issue, which was attacked by some hackers.

Luckily, this serious problem was dealt with in a timely manner.

This `hotfix` branch should be kept for the git log tracing, and also a good warning for the entire team of contributors, I would rather say.

# Changes include


# Breaking changes

It is not backward compatible, node version which `< v1.1.2` should update to version `>= v1.1.2`, and use the latest `genesis.json` in the [documentation](https://docs.dogechain.dog/docs/get-started/full-node-deployment#copy-genesisjson-to-etc-directory).

# Checklist


## Testing


### Manual tests

1. Set up a new network: 4 validators and `portland` hardfork block number set to `200` for fast verfication.
2. Use a `v1.1.1` node to sync.
    * It should be blocked on block number `200`.
3. Replace the node to `v1.1.2`.
    * It should be able to continue syncing.
4. Use some tool to send huge amount transactions to the network, set it up with `100 GWei` gas price.
5. Send a `50GWei` transaction by `MetaMask`. 
    * It should be pending for a long long time.
6. `Speed up` the transaction with `120GWei` gas price.
    * It should be included to some block very soon.
7. Enable the `WebSocket` of the rpc node, using `--enable-ws --json-rpc-block-range-limit=0` flag.
8. Connect to the WebSocket, create some block and log filters, get the changes.
    * The block range should not be limited.
9. Close the WebSocket connection, and watch the node log.
    * There should be `Closing WS connection with error` in logs, and all the timeout WebSocket filter should be removed only once.

Have tested manually, and all behaviors work like it is expected.

# Documentation update

https://github.com/dogechain-lab/dogechain-docs/commit/e2afea264c4b222b92de3461cbd0d6bfc540e720
